evalml/pipelines/components/transformers/samplers/base_sampler.py

```diff
@@ -57,7 +57,7 @@ def transform(self, X, y=None):
         X = infer_feature_types(X)
         if y is not None:
             y = infer_feature_types(y)
-        return X, y
+        return X, None
```

Comment:
I wrote up an explanation for why this fixes the issue in the PR description. Copying here for visibility:

> Because undersampler `transform` returns both X and y, our current `ComponentGraph` impl [will save the unmodified target returned from that method under the key `"Undersampler.y"`](https://github.com/alteryx/evalml/blob/main/evalml/pipelines/component_graph.py#L242) in the component features dict during evaluation. This doesn't cause problems during pipeline fit and predict, but when features are computed for each transformer [in `_fit_transform_features_helper`,](https://github.com/alteryx/evalml/blob/main/evalml/pipelines/component_graph.py#L168) that code erroneously grabs the `"Undersampler.y"` output **and appends it to the input of the next component**, which in automl and in our reproducer happens to be the estimator. This is why one of the stack traces at the bottom comes from the estimator and mentions too many features being present.

Reply:
Long-term, we'll need to update our component graph to be tolerant to this behavior. If a component `transform` happens to return a target, we should use that target. If it doesn't we should just use the last modified target. We have logic for this already in the main component graph method for fit/predict, but not yet in `_fit_transform_features_helper`. We can address this separately once this patch is merged and tested.